File: python_controller/src/python_controller/services/rabbitmq_communication_service.py
import json
import pika

from python_controller.messages import BaseMessage, get_message_from_json

class RabbitMQCommunicationService:
    class __RabbitMQCommunicationService:
        in_queue_name = "inzynierka_python"
        app_queue_name = "inzynierka_app"
        launcher_queue_name = "inzynierka_launcher"

        _message_class_name_json_key = "class_name"
        _message_body_json_key = "body"

        def __init__(self):
            self.connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost'))
            self.channel = self.connection.channel()

            self.channel.queue_declare(queue=self.in_queue_name)
            self.channel.queue_declare(queue=self.app_queue_name)
            self.channel.queue_declare(queue=self.launcher_queue_name)
            self.channel.queue_purge(queue=self.in_queue_name)
            self.channel.queue_purge(queue=self.app_queue_name)
            self.channel.queue_purge(queue=self.launcher_queue_name)

            self._message_subscriptions = {}
        
        def __str__(self):
            return repr(self) + self.val

        def _run_subscriptions(self, message):
            subscriptions = self._message_subscriptions.get(message.__class__.__name__)

            if subscriptions is None:
                return

            for subscription in subscriptions:
                subscription(message)

        def _callback(self, ch, method, properties, body):
            print("INFO: RabbitMQCommunicationService._callback: received message: " + body.decode())

            data = json.loads(body.decode())

            message_class_name = data.get(self._message_class_name_json_key)
            message_body = data.get(self._message_body_json_key)

            if (not message_class_name) or (not message_body):
                print ("ERROR: RabbitMQCommunicationService._callback: invalid json structure")
                return 

            message = get_message_from_json(message_class_name, message_body)

            if message is None:
                return
            
            self._run_subscriptions(message)

        def start(self, setup_finished_callback):
            self.channel.basic_consume(queue=self.in_queue_name, on_message_callback=self._callback, auto_ack=True)
            setup_finished_callback()
            self.channel.start_consuming()

        def subscribe(self, message_type: BaseMessage, callback_function):
            message_type_name = message_type.__name__
            subscriptions = self._message_subscriptions.get(message_type_name)

            if subscriptions is None:
                self._message_subscriptions[message_type_name] = [callback_function]
                return
            
            subscriptions.append(callback_function)

        def publish(self, message: BaseMessage):
            message_body_json = message.to_json()

            sent_json = json.dumps({
                self._message_class_name_json_key: message.__class__.__name__,
                self._message_body_json_key: message_body_json
            })

            print("publishing message: " + sent_json)
            self.channel.basic_publish(exchange='', routing_key=message.receiver, body=sent_json)

    instance = None
    def __init__(self, *args, **kwargs):
        if not RabbitMQCommunicationService.instance:
            RabbitMQCommunicationService.instance = RabbitMQCommunicationService.__RabbitMQCommunicationService()
    def __getattr__(self, name):
        return getattr(self.instance, name)

File: app/Inzynierka/CommunicationService/RabbitMQCommunicationService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

using RabbitMQ.Client;
using RabbitMQ.Client.Events;
using Windows.UI.Core;
using Windows.UI.Xaml;

using Inzynierka.CommunicationService.Messages;
using Inzynierka.Exceptions;
using Inzynierka.MainThreadDispatcher;
using Inzynierka.StateMessaging;

namespace Inzynierka.CommunicationService
{
    public sealed class RabbitMQCommunicationService
    {
        public static string PythonQueueName = "inzynierka_python";
        public static string IncomingQueueName = "inzynierka_app";
        public static string LauncherQueueName = "inzynierka_launcher";

        public bool Initialized = false;

        private static RabbitMQCommunicationService m_oInstance = null;
        private static readonly object m_oPadLock = new object();

        private ConnectionFactory Factory { get; set; }
        private IConnection Connection { get; set; }
        private IModel ConnectionModel { get; set; }

        private List<string> Queues = new List<string>();

        public event EventHandler<MessageReceivedEventArgs> MessageReceived;

        private Dictionary<string, List<Action<BaseIndication>>> MessageNameToCallbacks = new Dictionary<string, List<Action<BaseIndication>>>();

        public static RabbitMQCommunicationService Instance
        {
            get
            {
                lock (m_oPadLock)
                {
                    if (m_oInstance == null)
                    {
                        m_oInstance = new RabbitMQCommunicationService();
                    }
                    return m_oInstance;
                }
            }
        }

        private RabbitMQCommunicationService()
        {
        }

        public void Initialize()
        {
            Factory = new ConnectionFactory() { HostName = "localhost" };
            Connection = Factory.CreateConnection();
            ConnectionModel = Connection.CreateModel();
            DeclareOutgoingQueue(PythonQueueName);
            DeclareOutgoingQueue(LauncherQueueName);
            DeclareIncomingQueue(IncomingQueueName);

            foreach (string key in MessageDictionary.MessageClassNameToType.Keys)
            {
                MessageNameToCallbacks.Add(key, new List<Action<BaseIndication>>());
            }

            Initialized = true;
        }

        private void DeclareOutgoingQueue(string name)
        {
            if (string.IsNullOrWhiteSpace(name))
            {
                throw new ArgumentException("Queue name should consist of non-whitespace characters", nameof(name));
            }

            if (Queues.Contains(name))
            {
                throw new QueueAlreadyExistsException(name);
            }

            ConnectionModel.QueueDeclare(queue: name,
                durable: false,
                exclusive: false,
                autoDelete: false,
                arguments: null);
            Queues.Add(name);
            CleanQueue(name);
        }

        private void DeclareIncomingQueue(string name)
        {
            if (Queues.Contains(name))
            {
                throw new QueueAlreadyExistsException(name);
            }

            ConnectionModel.QueueDeclare(queue: name,
                durable: false,
                exclusive: false,
                autoDelete: false,
                arguments: null);
            Queues.Add(name);
            CleanQueue(name);

            var consumer = new EventingBasicConsumer(ConnectionModel);
            consumer.Received += Receiver;
            ConnectionModel.BasicConsume(queue: name,
                autoAck: true,
                consumer: consumer);

        }

        public void Send(BaseIndication message)
        {
            if (!Initialized)
            {
                throw new NotInitializedException();
            }

            if (!Queues.Contains(message.Receiver))
            {
                throw new QueueDoesntExistException();
            }

            var messageBodyJson = message.ToJson();
            var wrappingMessage = new WrappingMessage()
            {
                ClassName = message.GetType().Name,
                Body = messageBodyJson
            };
            var wrappingMessageJson = JsonSerializer.Serialize(wrappingMessage);
            var body = Encoding.UTF8.GetBytes(wrappingMessageJson);

            ConnectionModel.BasicPublish(exchange: "",
                routingKey: message.Receiver,
                basicProperties: null,
                body: body);
        }

        private async void Receiver(object model, BasicDeliverEventArgs ea)
        {
            var body = ea.Body.ToArray();
            string wrappingMessageJson = Encoding.ASCII.GetString(body);
            var outerMessage = JsonSerializer.Deserialize<WrappingMessage>(wrappingMessageJson);
            Type messageType;
            try
            {
                messageType = MessageDictionary.GetTypeFromClassName(outerMessage.ClassName);
            }
            catch(MessageTypeNotFoundException e)
            {
                StateMessagingService.Instance.SendInfoMessage(e.Message, 5000);

                return;
            }

            var deserializeMethod = typeof(JsonSerializer).GetMethod("Deserialize", new[] { typeof(string), typeof(JsonSerializerOptions) });
            var deserializeRef = deserializeMethod.MakeGenericMethod(messageType);
            var messageBody = deserializeRef.Invoke(null,
                new object[]
                {
                    outerMessage.Body,
                    new JsonSerializerOptions()
                    {
                        Converters = { new JsonStringEnumConverter() }
                    }
                });

            foreach (var callback in MessageNameToCallbacks[outerMessage.ClassName])
            {
                callback(messageBody as BaseIndication);
            }

            await MainThreadDispatcherService.MarshalToMainThreadAsync(() => MessageReceived?.Invoke(this, new MessageReceivedEventArgs(messageBody)));
        }

        public void Subscribe(string messageName, Action<BaseIndication> callback)
        {
            if (!MessageDictionary.MessageClassNameToType.Keys.Contains(messageName))
            {
                throw new MessageTypeNotFoundException();
            }

            MessageNameToCallbacks[messageName].Add(callback);
        }

        public void CleanQueue(string queueName)
        {
            ConnectionModel.QueuePurge(queueName);
        }

    }
}
